### PR TITLE
Add size support to `top_metrics`

### PR DESCRIPTION
**Pull request by nik9000**
This adds support for returning the top "n" metrics instead of just the
very top.

Relates to #51813

### CODE REVIEW

**Comment by nik9000** [comment]
@not-napoleon and @polyfractal, about half of this is plumbing. The other half is the fancy min-heap. The third half is testing.

**Comment by nik9000** [comment]
This will need #52575 to be merged first.

**Comment by nik9000** [comment]
> This will need #52575 to be merged first.

And that is in! I think this is ready for review when anyone has time.

**Comment by nik9000** [comment]
I'm going to resolve the merge conflict and add an integration test for the new parameter `top_metrics_max_bucket`.

**Review by not-napoleon — APPROVED** [review]
I think this looks good overall.  I left several nits and requests for clarification, but (other than the typo fix) I don't think any are strictly necessary.

File: server/src/main/java/org/elasticsearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java [hunk 1]
```diff
@@ -104,23 +104,29 @@ protected NumericDocValues getNumericDocValues(LeafReaderContext context, String
     }
 
     @Override
-    public BucketedSort newBucketedSort(BigArrays bigArrays, SortOrder sortOrder, DocValueFormat format) {
-        return new BucketedSort.ForLongs(bigArrays, sortOrder, format) {
+    public BucketedSort newBucketedSort(BigArrays bigArrays, SortOrder sortOrder, DocValueFormat format,
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java]: They differ by the primitive types they operate on. I don't think they *can* share very much more than they do. But I can certainly look into sharing more.

File: server/src/main/java/org/elasticsearch/search/sort/BucketedSort.java [hunk 2]
```diff
@@ -86,126 +233,305 @@ public final SortValue getValue(long bucket) {
...
+     * @param rootIndex index of the start of the bucket
+     * @param parent Index within the bucket of the parent to check.
+     *               For example, 0 is the "root". 
+     */
+    private void downHeap(long rootIndex, int parent) {
+        while (true) {
+            long parentIndex = rootIndex + parent;
+            int worst = parent;
+            long worstIndex = parentIndex;
+            int left = parent * 2 + 1;
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/sort/BucketedSort.java]: :+1:

File: server/src/main/java/org/elasticsearch/search/sort/BucketedSort.java [hunk 3]
```diff
@@ -86,126 +233,305 @@ public final SortValue getValue(long bucket) {
...
+     * Heapify a bucket who's entries are in random order.
+     * <p>
+     * While this *looks* like it could easily be {@code O(n * log n)}, it is
+     * a fairly well studied algorithm attributed to Floyd there's
+     * been a bunch of work that puts this at {@code O(n)}, close to 1.88n worst
+     * case.
+     * </p>
+     * <ul>
+     * <li>Hayward, Ryan; McDiarmid, Colin (1991).  
+     * <a href="https://web.archive.org/web/20160205023201/http://www.stats.ox.ac.uk/__data/assets/pdf_file/0015/4173/heapbuildjalg.pdf">
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/sort/BucketedSort.java]: :+1:

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java [hunk 4]
```diff
@@ -17,52 +19,53 @@
...
      */
     public InternalTopMetrics(StreamInput in) throws IOException {
         super(in);
-        sortFormat = in.readNamedWriteable(DocValueFormat.class);
         sortOrder = SortOrder.readFromStream(in);
-        sortValue = in.readOptionalNamedWriteable(SortValue.class);
         metricName = in.readString();
-        metricValue = in.readDouble();
+        size = in.readVInt();
+        topMetrics = in.readList(TopMetric::new);
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java]: We haven't released it, yeah. So, indeed, breaking serialization is "safe". I have to disable BWC tests until I backport it, but other than that this should be ok.

File: server/src/main/java/org/elasticsearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java [hunk 5]
```diff
@@ -96,23 +96,29 @@ public void setScorer(Scorable scorer) {
...
+                    private double docValue;
 
                     @Override
                     protected boolean advanceExact(int doc) throws IOException {
-                        return values.advanceExact(doc);
+                        if (values.advanceExact(doc)) {
+                            docValue = values.doubleValue();
+                            return true;
+                        }
+                        return false;
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java]: I got it in my head that I wanted `docValue` be "fast" and the underlying Lucene iterators aren't. I don't recall why I wanted that. Let me try to suss out what past-Nik was thinking.

[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/fielddata/fieldcomparator/DoubleValuesComparatorSource.java]: OK! I've figured it out. I've also pushed javadoc explaining it, but I'll say it here too: when you are in heap mode we call `docValue()` twice, once to check if the doc is competitive, and once to load the value into the heap. The underlying doc values code may or may not be fast. It looks like it most cases they are indeed fast, but this is a little bit of extra paranoia.

File: server/src/main/java/org/elasticsearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/index/fielddata/fieldcomparator/LongValuesComparatorSource.java]: That's fair.  I didn't spend a lot of time thinking about it other than to see that they looked really similar.  If it's necessary, it's necessary.

**Comment by not-napoleon** [comment]
Changes look great, thanks for doing them!

**Comment by nik9000** [comment]
> Changes look great, thanks for doing them!

Thanks for the review!

**Pull request merged**